tutorial/Execution.md

```diff
@@ -0,0 +1,412 @@
+
+```
+object Execution {
+  final def from[T](value: => T): Execution[U] = {
+    def newDoSomething(config: Config, mode: Mode) = value
+    Execution(newDoSomething)
+  }
+
+  // ... other previously defined methods ...
+}
+```
+
+Why is it `value: => T` and not `value: T`? This scala syntax here `=>` is "call by name" and is essentially receiving a function of 0 parameters. Check out the scala documentation for more.
+
+At this point, we can write an `Execution` that reads from a service and uses its result to do something:
+
+```scala
+val config: Config = // ...
+val mode: Mode = // ...
+val service: String => String = // ...
+
+val callOutToService = Execution.from {
+  service("eat galaxies")
+}
+
+val reversedServiceResult = callOutToService.map { serviceResult =>
+  serviceResult.reverse
+}
+
+reversedServiceResult.run(config, mode)
+```
+
+### Running a Scalding job
+
+Writing a single Scalding job with `TypedPipe` and `Grouped` will not be covered in this tutorial; however, in order to run Scalding jobs with `Execution`, we need to be familiar with with 2 methods on `TypedPipe`: `writeExecution` and `toIterableExecution`:
+
+```scala
+val config: Config = // ...
+val mode: Mode = // ...
+
+val source: TypedSource[Int] = // ...
+val transform: Int => String = // ...
+val sink: TypedSink[String] = // ...
+
+val scaldingJob: Execution[Unit] = TypedPipe
+  .from(source)
+  .map(transform)
+  .writeExecution(sink)
+
+scaldingJob.run(config, mode)
+
+val scaldingJobResults: Execution[Iterable[String]] = TypedPipe
+  .from(source)
+  .map(transform)
+  .toIterableExecution
+
+scaldingJobResults
+  .map { results => results.foreach(println) }
+  .run(config, mode)
+```
+
+The `TypedPipe` method `writeExecution` parallels the normal `write` method on `TypedPipe`, but instead of planning the result inside of a Scalding `Job` class, `writeExecution` yields a plan to run that Scalding job as an `Execution`. The type parameter is `Unit` because the purpose of the `Execution` created from the `TypedPipe` is to have side-effects by writing to the `TypedSink` argument.
+
+The `TypedPipe` method `toIterableExecution` creates an `Execution` plan to expose the output of the `TypedPipe` in the local environment. This is useful if we need to use the output of a Scalding job to decide what to do next. Note that there's no guarantee on the order of the data and for large datasets, the length of the `Iterable` could be significant.
+
+### Flatmap
+
+So far, we've seen how to examine the `Config`, `Mode`, lift a value into the Execution environment, and run a Scalding job. We've alluded to the idea that we can use the output of one `Execution` to plan another, but haven't talked about how to do that. One may be familiar with the method `flatMap` on various collections in scala, just like `map`. `flatMap` acts differently depending on the type of collection:
+
+```scala
+val toThree        = Seq.range(0, 3)                    // Seq(0, 1, 2)
```

Comment:
Seq.flatMap is kind of confusing compared to Option.flatMap, which works more like Execution/Future .flatMap. Maybe stick with Option?